avocado/plugins/assets.py

```diff
@@ -0,0 +1,179 @@
+        Visit calls and build call string if call matches the pattern
+        :param node: (Sub)Tree root to start the parse
+        :type node: ast.*
+        """
+        if isinstance(node.func, ast.Attribute):
+            if self.pattern in node.func.attr:
+                call = self._build_call(node)
+                if call is not None:
+                    self.calls.append(call)
+    # visit_call()
+
+    def fetcher(self):
+        """
+        Build an intrumented avocado test with all fetch_asset calls and
+        execute it.
+        """
+        # build raw code from list of calls, if available
+        if len(self.calls) > 0:
+            rcalls = ""
+            for call in self.calls:
+                rcalls = rcalls + call
+        else:
+            LOG_UI.debug('No supported assets found to fetch from %s'
+                         % self.file_name)
+            return
+
+        # temporary test source file
+        fetch_test = tempfile.NamedTemporaryFile(suffix=".py")
+        fetch_test.write(TEST_TEMPLATE.format(fetch_assets=rcalls).encode())
+        fetch_test.flush()
+
+        # run teporary test
+        LOG_UI.debug('Fetching assets from %s' % self.file_name)
+        self.result = process.run("avocado run --json - %s" % fetch_test.name,
+                                  ignore_status=True)
+        fetch_test.close()
+
+        # check if test ran without problems
+        fetch_test_stdout = json.loads(self.result.stdout)
+        if (self.result.exit_status != 0
+                or fetch_test_stdout['pass'] != fetch_test_stdout['total']):
+            LOG_UI.error('%s Check %s for more details.' % (
+                fetch_test_stdout['tests'][0]['fail_reason'],
+                fetch_test_stdout['tests'][0]['logfile']))
+    # fetcher()
+# FetchAssetHandler()
+
+
+class Assets(CLICmd):
+    """
+    Implements the avocado 'assets' subcommand
+    """
+    name = 'assets'
+    description = 'Manage assets'
+
+    def configure(self, parser):
+        """
+        Add the subparser for the assets action.
+
+        :param parser: Main test runner parser.
+        """
+        parser = super(Assets, self).configure(parser)
+
+        subcommands = parser.add_subparsers(dest='assets_subcommand')
+        subcommands.required = True
+
+        fetch_subcommand_parser = subcommands.add_parser('fetch',
+                                help='Fetch assets from test source'\
+                                ' or config file if it\'s not already in the'\
+                                ' cache')
+        fetch_subcommand_parser.add_argument('references', type=str,
+                                             default=[], nargs='*',
+                                             metavar='TEST_REFERENCE',
+                                             help='List of test references')
+    # configure()
+
+    def run(self, config):
+        subcommand = config.get('assets_subcommand')
+
+        if subcommand == 'fetch':
```

Comment:
I know the argparse should not let the `assets_subcommand` to be different than supported ones, but just to be safe I'd recommend adding `else` with error message about unsupported subcommand. That way we'd spot typos easier...

Reply:
There is already an error message for unsupported subcommands by default:

```
$ avocado assets fetche
usage: avocado assets [-h] {fetch} ...

positional arguments:
  {fetch}
    fetch     Fetch assets from test source or config file if it's not already
              in the cache

optional arguments:
  -h, --help  show this help message and exit

avocado assets: error: argument assets_subcommand: invalid choice: 'fetche' (choose from 'fetch')
```

That is the reason I have not added another check, as I would expect other commands from avocado follow the same to keep subcommand error messages standard.

Reply:
Sure, in ideal world yes, but consider hunting down a "typo"... We already use `if/elif/...`, let's add the 2 lines of `else: raise ...` with really simple error message as in 99.99% cases it won't hit the user. But it might...

Reply:
Just a heads up, I added the `else` statement in the v2 implementation, but the parser handles unknown subcommands into the `config` method, so the `else` won't make any effect here.